### DISCUSSION TITLE
Correlated Scalar Subquery Optimisation

### DISCUSSION THREAD

**Discussion by jgrant-the-giant** [duckdb/duckdb#14932] [General]
Hello, I'm seeking to understand if there's any (long-term) potential for optimisation within the engine when selecting correlated scalar sub queries.

For example:

```
from thetable a
select 
a.dimension, 
(select count(*) from thetable b where a.dimension = b.dimension) total
(select count(*) from thetable b where a.dimension = b.dimension and b.filtered = 1) filtered_total
group by a.dimension
```

being logically equivalent to but slower than

```
from thetable
select dimension, 
count(*) total, 
sum(case when filtered=1 then 1 else 0 end) as filtered_total
group by dimension
```

Thank you!

**Comment by soerenwolfers** [comment]
Do you have a real-world example where the former is more natural to write than the latter and the slowness of the former has caught you out?

If yes, that might make it more likely that effort is devoted to this.

By the way, you can use 
```sql
count() filter (filtered=1)
``` 
or, on recent versions, 
```sql
sum(filtered=1)
```
instead of the verbose
```sql 
sum(case when filtered=1 then 1 else 0 end)
```

**Comment by jgrant-the-giant** [comment]
Thank you for the syntax advice, that's very elegant. Apologies in advance for the long post..

===PRELUDE===

I have many such queries, as I've written a python package that generates them in an attempt to explore an OSS replacement for MS's Data Analysis Expressions (DAX) language. All with the intention of making it easier to build nested calculations that respond to user filters and other dynamic contexts. I'm not here to promote my tool, which needs to be rebuilt, but I would like to suggest that there's potential in the overall approach.

The DAX language features composable "measures" that are essentially the scalar subqueries mentioned. In my code, filters are sustained through composition using equi-join clauses: "a.dimension=b.dimension". 

The structure feels very powerful, a kind of generalised aggregate that can reach across different schemas through equi-join semantics. It makes analysis of snowstorm schemas (groups of related snowflake schemas) much easier. DAX context modifiers like ALL(), which can clear a filter context on one or more columns, can be achieved simply by omitting the relevant joins. Lookups become straightforward and it's easy to manage cardinality differences.

The "naturality" is in a vaguely mathematical sense. I would counter that it's very "unnatural" for selected scalar subqueries not to aquire the filters generated implicitly in a group by expression. The slowness I have noticed by making comparisons of like-for-like queries in PowerBI - DuckDB evaluation is linear in the number of these selected scalar subqueries, whereas Vertipac is very sub-linear. What I'm doing is much more nested but there are clearly gains to make by considering the example I gave originally.

===AN EXAMPLE===

An example which gives the top 3 most commonly reported results for my data by month, string concatenated together so that the grain is by month. 

- The object C references the FACT_TABLE and knows its schema
- If C receives a measure on a schema with common dimensions it generates the equi-joins automatically.
- This includes measures from FACT_TABLE itself.

The python code:
```
most_answered_questions = C.string_column_sample(
    'string_concatenated_results', 
    ['question', 'response'],
    sort_by=C.size() )

agg_by_month = C(
    outer=most_answered_questions,
    dim='month_of_wc_monday', 
    order_by='all'
)
```

The resulting SQL:
```
select month_of_wc_monday,

  (with __VIEW__1_0 as
     (select question,
             response,

        (select concat_ws(', ', question, response) __the__words
         from FACT_TABLE __ALIAS__1_1
         where (__ALIAS__1_1.month_of_wc_monday is not distinct
                from __ALIAS__0_0.month_of_wc_monday)
           and (__ALIAS__1_1.question is not distinct
                from __ALIAS__0_1.question)
           and (__ALIAS__1_1.response is not distinct
                from __ALIAS__0_1.response)
         group by question,
                  response) __the__words,

        (select count(*) __string_sample_order
         from FACT_TABLE __ALIAS__1_1
         where (__ALIAS__1_1.month_of_wc_monday is not distinct
                from __ALIAS__0_0.month_of_wc_monday)
           and (__ALIAS__1_1.question is not distinct
                from __ALIAS__0_1.question)
           and (__ALIAS__1_1.response is not distinct
                from __ALIAS__0_1.response) ) __string_sample_order
      from FACT_TABLE __ALIAS__0_1
      where (__ALIAS__0_1.month_of_wc_monday is not distinct
             from __ALIAS__0_0.month_of_wc_monday)
      group by question,
               response
      order by __string_sample_order desc
      limit 3) select string_agg(__the__words, ' --- ') string_concatenated_results
   from __VIEW__1_0 __ALIAS__1_0) string_concatenated_results

from FACT_TABLE __ALIAS__0_0
group by month_of_wc_monday
order by all
```

This yields results like

| **month_of_wc_monday** |	**string_concatenated_results** |
| 2022-11 Nov | "General Comment, No Issues --- Store Contact, Manager --- Some Other Important Question, Most Usual Response" |

===FINALLY===

I know I'm synthesising "elementary" SQL and expecting the query optimiser to make up the difference. It's an unfair ask. But I think DuckDB can approach Power BI's performance if it's possible to (even partially) decorrelate these queries, even in the simplest cases like the original example. I'm focused on that more simple example because it's genuinely like-for-like with an optimal version.

TL;DR It's possible to investigate optimising this type of SQL structure as a means to compete with MS Power BI and the (marvellous) Vertipaq engine, for generating complex, dynamic SQL in user facing applications.
